Content.Server/Mind/MindSystem.cs

```diff
@@ -502,87 +527,126 @@ public bool TryRemoveObjective(Mind mind, int index)
     /// </summary>
-    /// <param name="mind">The mind to remove the role from.</param>
-    /// <param name="role">The type of the role to remove.</param>
+    /// <param name="mindId">The mind to remove the role from.</param>
+    /// <typeparam name="T">The type of the role to remove.</typeparam>
     /// <exception cref="ArgumentException">
     ///     Thrown if we do not have this role.
     /// </exception>
-    public void RemoveRole(Mind mind, Role role)
+    public void RemoveRole<T>(EntityUid mindId) where T : Component
     {
-        if (!mind.Roles.Contains(role))
+        if (!RemComp<T>(mindId))
         {
-            throw new ArgumentException($"We do not have this role: {role}");
+            throw new ArgumentException($"We do not have this role: {typeof(T)}");
         }
 
-        mind.Roles.Remove(role);
-
-        var message = new RoleRemovedEvent(mind, role);
+        var mind = Comp<MindComponent>(mindId);
+        var antagonist = _roles.IsAntagonistRole<T>();
+        var message = new RoleRemovedEvent(mindId, mind, antagonist);
 
         if (mind.OwnedEntity != null)
         {
             RaiseLocalEvent(mind.OwnedEntity.Value, message, true);
         }
         _adminLogger.Add(LogType.Mind, LogImpact.Low,
-            $"'{role.Name}' removed from mind of {MindOwnerLoggingString(mind)}");
+            $"'Role {typeof(T).Name}' removed from mind of {MindOwnerLoggingString(mind)}");
+    }
+
+    public void TryRemoveRole<T>(EntityUid mindId) where T : Component
+    {
+        if (HasRole<T>(mindId))
+            RemoveRole<T>(mindId);
     }
 
-    public bool HasRole<T>(Mind mind) where T : Role
+    public bool HasRole<T>(EntityUid mindId) where T : Component
     {
-        return mind.Roles.Any(role => role is T);
+        return HasComp<T>(mindId);
     }
 
-    public bool TryGetSession(Mind mind, [NotNullWhen(true)] out IPlayerSession? session)
+    public bool TryGetSession(EntityUid? mindId, [NotNullWhen(true)] out IPlayerSession? session)
     {
-        return (session = mind.Session) != null;
+        session = null;
+        return TryComp(mindId, out MindComponent? mind) && (session = mind.Session) != null;
     }
 
     /// <summary>
     /// Gets a mind from uid and/or MindContainerComponent. Used for null checks.
     /// </summary>
     /// <param name="uid">Entity UID that owns the mind.</param>
+    /// <param name="mindId">The mind id.</param>
     /// <param name="mind">The returned mind.</param>
-    /// <param name="mindContainerComponent">Mind component on <paramref name="uid"/> to get the mind from.</param>
+    /// <param name="container">Mind component on <paramref name="uid"/> to get the mind from.</param>
     /// <returns>True if mind found. False if not.</returns>
-    public bool TryGetMind(EntityUid uid, [NotNullWhen(true)] out Mind? mind, MindContainerComponent? mindContainerComponent = null)
+    public bool TryGetMind(
+        EntityUid uid,
+        out EntityUid mindId,
```

Comment:
Is there a reason why this isn't nullable?

Reply:
It makes it harder to use the API with value types, same reason why Dictionary doesn't explicitly make the out parameter in TryGetValue nullable, as it would force you to do .Value on valuetypes, even when you know that they are not null.

Reply:
I'm pretty sure it doesn't force .Value if you put NotNullWhen and properly use the TryGet.

Reply:
![image](https://github.com/space-wizards/space-station-14/assets/10968691/911c1553-b365-4a33-ac0e-d819380a0225)
![image](https://github.com/space-wizards/space-station-14/assets/10968691/6e10bb34-6cca-42e0-b1cb-996329fe2c1e)